### PR TITLE
Remove wait from poll package

### PR DESCRIPTION
**Pull request by pacevedom**
Remove wait calls from poll package. Reconcile loops were stuck with 30s waiting time per resource, and if conditions were not true it bailed out with an error to retrigger a new reconciliation loop.
This change checks for immediate condition or else return error to allow other CRs to go through and reconcile, while also shortening a bit the full reconcile ok status.

### CODE REVIEW

**Comment by zvonkok** [comment]
SRO will not wait for every resource 30s ... Most of the resources are done in a couple of seconds. If a resource needs more than 30sec to deploy this is indeed something that needs some attention. The retry interval is 5 sec. If the reconciled resource is a dependency then you would need to wait for it anyway. This is also used to minimize the load on the API.

**Comment by zvonkok** [comment]
If you have a lot of nodes you would "bombard" the API server with get requests  I would suggest we turn down the retry and timeout values if this is a concern but would not remove it completely.

**Comment by pacevedom** [comment]
Hey Zvonko,

If the conditions are not met for a given resource (be that availability/unavailability) the waiting hits 30s before bailing out. If we imagine a `BuildConfig` then this cycle would repeat several times. Also, the use of `wait.Poll` instead of `wait.PollImmediate` adds 5s to each resource conditions check.
If we are going to fail the reconcile after retrying, then there is no big difference between this method and retriggering a new reconcile loop, except for the fact that we block the operator. There is only 1 concurrent reconcile, so the longer we spend time with a single `SpecialResource`, the longer we delay others, as they start to queue up.

You mention bombarding the API server and you got a strong point there, but if we look at the full reconcile loop (even after this change) we see there is still significant wait time due to other operations in the loop. If we compare the number of reconcile loops executed with `ping-pong` between old and new change it is 5 vs 7 (worth noting for comparison between # of loops that wait-less approach is between 1 and 2 minutes faster). In the wait approach there are several API calls per loop, while the wait-less approach only makes 1 call per resource per loop.
If API consumption becomes a problem we can always change to a watch strategy instead of getting info directly in every loop. This would also lead to requeue reconciles when not ready instead of returning errors, which are transient (and sometimes expected).

Thanks!

**Comment by zvonkok** [comment]
We should not wait for resources that do not have the wait annotation, but there are some resource where this is needed per default. 

If a Pod mounts a ConfigMap or Secret you should wait until those resources are available otherwise 
you break the Pod, Veritas had this very problem, because SRO was too fast creating those resources. Pod was ready and the CM or Secret were still in "creation" phase. 

Why remove something when we know we might need it in the future?

Some states need to be done before the next state is executed. Perfect example you have given is the `BuildConfig` I need the kernel module before I am attempting lets say create VFs or configure some sys values or need build artifacts for a subsequent SR. I need to wait, yes we block the operator but for a good reason.

Currently, as of now, other operations take some time, but moving forward and optimizing those may get us in the situation where we need to enable wait times again. 

I agree that we should reconcile another SR if possible if we wait "too" long for a resource.

**Comment by zvonkok** [comment]
@pacevedom What do you think of adding "performance metrics" to SRO and reporting them as well in the Status or Prom? How long did each state took, chart took, or special resource?

**Comment by zvonkok** [comment]
This could be used also for regression testing, if changes introduced too much "overhead".

**Comment by pacevedom** [comment]
> We should not wait for resources that do not have the wait annotation, but there are some resource where this is needed per default.

Reading this makes me think I didnt explain myself properly before. What I am suggesting to remove is not the entire wait, only the use of `wait` package in the code in order to not stall the reconcile loop. The annotations for waiting are still there, this is not a behavior change, only a speedup in checking if I need to reconcile something else.

> If a Pod mounts a ConfigMap or Secret you should wait until those resources are available otherwise you break the Pod, Veritas had this very problem, because SRO was too fast creating those resources. Pod was ready and the CM or Secret were still in "creation" phase.

Indeed, and these will keep waiting for the status just like before.

> Why remove something when we know we might need it in the future?

I agree on keeping the wait because that is one of the main points of SRO, but not so much on the `wait` package. My change suggests only to not block the main loop when we are just waiting and could check later.

> Some states need to be done before the next state is executed. Perfect example you have given is the `BuildConfig` I need the kernel module before I am attempting lets say create VFs or configure some sys values or need build artifacts for a subsequent SR. I need to wait, yes we block the operator but for a good reason.

But arent such resources marked with the wait annotation? If so, there is no problem, right?

> Currently, as of now, other operations take some time, but moving forward and optimizing those may get us in the situation where we need to enable wait times again.
> 
> I agree that we should reconcile another SR if possible if we wait "too" long for a resource.

This is what the change aims. Look if everything that is needed is ready, and if not bail out and make way for another reconcile loop. This helps paralelism when having several SRs. Or even helps with lowering down a single SR, as can be seen in the tests. The e2e tests are now 2-3 minutes faster.

> @pacevedom What do you think of adding "performance metrics" to SRO and reporting them as well in the Status or Prom? How long did each state took, chart took, or special resource?

I always thought that having as many metrics as possible helps once you are in ops, but I guess it depends on the cardinality of what we are pushing to prometheus. Maybe start small and add more details if needed? I would only do Prometheus, though, the status should be for something else to me.

> This could be used also for regression testing, if changes introduced too much "overhead".

We would have some insights of degrading performance over time, for example? Metrics are always a good idea to me.

**Comment by qbarrand** [comment]
Hello all,

> If a Pod mounts a ConfigMap or Secret you should wait until those resources are available otherwise you break the Pod, Veritas had this very problem, because SRO was too fast creating those resources. Pod was ready and the CM or Secret were still in "creation" phase.

Won't Kubernetes try to start the Pod until the ConfigMap becomes available? Whis is custom logic required there?

**Comment by zvonkok** [comment]
No, it won't. That is the whole point of this code. This was not added by "accident". Veritas had this problem with Secrets, some other occasions failed with ConfigMaps.

**Comment by qbarrand** [comment]
I can't reproduce what you're saying. Let's have a look at a simple Pod:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: cm-mount
spec:
  containers:
  - name: bash
    image: bash
    command: [sleep, infinity]
    volumeMounts:
    - name: cm-dependency
      mountPath: /mnt/cm
      readOnly: true
  volumes:
  - name: cm-dependency
    configMap:
      name: dependency
```
Creating this will result in the Pod being in the `ContainerCreating` state, with the following events:
```
Events:
  Type     Reason       Age                  From               Message
  ----     ------       ----                 ----               -------
  Normal   Scheduled    20m                  default-scheduler  Successfully assigned default/cm-mount to gke-quba-test-default-pool-c8be7fab-d6w4
  Warning  FailedMount  103s (x17 over 20m)  kubelet            MountVolume.SetUp failed for volume "cm-dependency" : configmap "dependency" not found
  Warning  FailedMount  3s (x9 over 18m)     kubelet            Unable to attach or mount volumes: unmounted volumes=[cm-dependency], unattached volumes=[cm-dependency kube-api-access-strgr]: timed out waiting for the condition
```
As expected, the Pod cannot be created because the `dependency` ConfigMap does not exist. Let's add it:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: dependency
data: {}
```
The Pod is then able to start:
```json
[
  {
    "lastProbeTime": null,
    "lastTransitionTime": "2022-03-02T08:32:56Z",
    "status": "True",
    "type": "Initialized"
  },
  {
    "lastProbeTime": null,
    "lastTransitionTime": "2022-03-02T08:55:35Z",
    "status": "True",
    "type": "Ready"
  },
  {
    "lastProbeTime": null,
    "lastTransitionTime": "2022-03-02T08:55:35Z",
    "status": "True",
    "type": "ContainersReady"
  },
  {
    "lastProbeTime": null,
    "lastTransitionTime": "2022-03-02T08:32:56Z",
    "status": "True",
    "type": "PodScheduled"
  }
]
```
Same goes with a Secret.
Could you please share reproduction steps?

**Comment by pacevedom** [comment]
Tried using `ConfigMap` and `Secret` without mounting them, as environment variables references and behavior is the same as the one @qbarrand pointed out. What are the triggering conditions to make it fail?

Aside from that, what do you think about removing the wait package calls? As this was the intention of this PR. Tests show a significant speedup.

**Comment by zvonkok** [comment]
@qbarrand @pacevedom my last comment had some "poor" wording. Let me think about the situation where we needed to wait for the CM or Secret to come up with the Pod even if the RestartPolicy was set to OnFailure or Always. A Pod that has RestartPolicy set to Never may fail if we do not wait for the CM or Secret that may be populated by a webhook that takes longer.

**Comment by zvonkok** [comment]
@pacevedom Let me try out your PR with my charts that have dependencies and longer wait times to see how this behaves. I get the point, the wait time would be implicitly the reconcile loop until it gets to the very same resource.

**Comment by zvonkok** [comment]
@qbarrand @pacevedom We should also start to rename the annotations from openshift to k8s.

**Comment by qbarrand** [comment]
> A Pod that has RestartPolicy set to Never may fail if we do not wait for the CM or Secret that may be populated by a webhook that takes longer.

I can see this happening indeed. Would [Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/) be a solution there?

File: pkg/poll/poll.go [hunk 1]
```diff
@@ -454,25 +432,19 @@ func (p *pollActions) forResourceFullAvailability(ctx context.Context, obj *unst
...
-			p.log.Info("Resource available ", "Kind", obj.GetKind()+": "+obj.GetNamespace()+"/"+obj.GetName())
-			return true, nil
-		}
-		p.log.Info("Waiting for availability of ", "Kind", obj.GetKind()+": "+obj.GetNamespace()+"/"+obj.GetName())
-		return false, nil
-	})
+	if fnd {
+		return nil
+	}
+	return fmt.Errorf("resource %s:%s/%s not available.", obj.GetKind(), obj.GetNamespace(), obj.GetName())
```
Comment (zvonkok) [inline pkg/poll/poll.go:447]:
```suggestion
	return fmt.Errorf("resource %s:%s/%s not available", obj.GetKind(), obj.GetNamespace(), obj.GetName())
```

**Comment by zvonkok** [comment]
Ok tested this PR and I am not satisfied with the result now I also remember why the wait was added in the first place. 
Without the wait, it takes far longer than with the wait. 

The problem is that the controller runtime is doing an exponential backoff and for every loop that the SR is not ready the
wait time gets longer and longer. Since some resources can take several minutes the backoff time is getting larger and larger and may even exceed the resource creation time. Each `|` is an event, and '-' is a time slice.  If the resource is not ready you get a dozen of reconcile events that lead to increased backoff time. 

```
                     CREATE                                                                                           READY
RES TIMELINE:        |-------------------------------------------------------------------------------------------|
RECONCILE TIMELINE:  ||||||||||-|--|----|--------|----------------|----------------------------------------|--------------------------------------------------------------------------------| 
```
If you only have resource that can be created fast it might be a good idea, but if you need to wait for a resource this will lead to longer wait times. Besides that log size is "dramatically" increased.

**Comment by pacevedom** [comment]
> @qbarrand @pacevedom my last comment had some "poor" wording. Let me think about the situation where we needed to wait for the CM or Secret to come up with the Pod even if the RestartPolicy was set to OnFailure or Always. A Pod that has RestartPolicy set to Never may fail if we do not wait for the CM or Secret that may be populated by a webhook that takes longer.

Might be wrong here, but isnt `restartPolicy` supposed to w+ork after the container has already started? If there is a missing dependency (like configmaps or secrets) it wont even start because requirements are not met. I checked doing so with secrets and pod didnt start until secret was available.


> Ok tested this PR and I am not satisfied with the result now I also remember why the wait was added in the first place. Without the wait, it takes far longer than with the wait.
> 
> The problem is that the controller runtime is doing an exponential backoff and for every loop that the SR is not ready the wait time gets longer and longer. Since some resources can take several minutes the backoff time is getting larger and larger and may even exceed the resource creation time. Each `|` is an event, and '-' is a time slice. If the resource is not ready you get a dozen of reconcile events that lead to increased backoff time.
> 
> ```
>                      CREATE                                                                                           READY
> RES TIMELINE:        |-------------------------------------------------------------------------------------------|
> RECONCILE TIMELINE:  ||||||||||-|--|----|--------|----------------|----------------------------------------|--------------------------------------------------------------------------------| 
> ```

Thanks! The backoff we are using is the default one: https://github.com/kubernetes-sigs/controller-runtime/blob/v0.11.1/pkg/controller/controller.go#L119-L121. 
This behavior can be overcome by either configuring custom rate limiters, or using `requeueAfter`. The first option would render uniform behavior over all resources, while the second one would allow custom waiting times for next reconcile loop.
What is common for both is the ability to serve other SRs while waiting for the one in progress.

With current approach we get 5s per resource when its ready, and up to 30 when its not. Imagine we have 10 resources (which is not a huge number) and the last one isnt ready. Thats 5*9+30=75s of waiting on resources. And that is just on the waiting side of resources, until we reach this point it takes some time too (an additional ~20s, at least). We are blocking the loop for a minute and a half each time we try to reconcile.

> If you only have resource that can be created fast it might be a good idea, but if you need to wait for a resource this will lead to longer wait times. Besides that log size is "dramatically" increased.

This is true, logs are quite verbose now. Some logs could well be either removed or squashed/merged with others, as they are pretty similar or dont give too much information.

**Comment by zvonkok** [comment]
EnqueueRequestForOwner will  trigger a reconcilitation regardles the requeueAfter. You might still get into too many reconciles.

Since SRO is doing the SRs one by one in the order of creation, how do you know which resource is stuck? What if your first one is stuck? The others were not considered in that case. 

If you change the behavior that the SR is being reconciled that created the event you essentially gained nothing. 

How is this going to work with helm-deps? Since those aren't SRs anymore you cannot treat them in some specific order.

**Comment by zvonkok** [comment]
If you have two SRs that have resources that are created first that need a long time, you would create too many events and get in the situation that I explained before.

**Comment by pacevedom** [comment]
> EnqueueRequestForOwner will trigger a reconcilitation regardles the requeueAfter. You might still get into too many reconciles.

Those come from the `doWatch`, right? Anyway those only happen when the owned resources change, and given the list that we have I wouldnt want to skip any, as it might mean you are done with some stage of the reconcile. Am I wrong?

> Since SRO is doing the SRs one by one in the order of creation, how do you know which resource is stuck? What if your first one is stuck? The others were not considered in that case.

Not sure I am following. Do you mean resources in a recipe for a given SR? As mentioned, this is not changing the waiting order nor the execution order, it reduces waiting time allowing for more reconciles to happen. This means faster reaction time.

> If you change the behavior that the SR is being reconciled that created the event you essentially gained nothing.

Thats the key, this PR does not change behavior as it is functionally equivalent. It swaps a small number of long reconciles for a higher number of shorter reconciles. Which, in turn, means you are able to serve more reconciles for any other SRs, plus finish faster with a single one as you minimize waiting slack. You can have a look at the CI, execution time has been reduced significantly.

> How is this going to work with helm-deps? Since those aren't SRs anymore you cannot treat them in some specific order.

I would say it relates to helm dependency handling, not waiting, or is it related to waiting?

> If you have two SRs that have resources that are created first that need a long time, you would create too many events and get in the situation that I explained before.

It does create more reconciliation loops, but is this a bad thing? It allows all SRs to progress faster. If the situation you refer to is the one about backoffs, that is easily configurable by having the exponential delay capped (which I included). If the calls to API server are too much we can also configure a bucket as a rate limiter.

**Comment by zvonkok** [comment]
Let me check with these new settings and get back to you with your questions and my poor explaining. :)

**Pull request closed without merging**